managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java

```diff
@@ -178,8 +178,21 @@ public void addComplete(int rc, final LedgerHandle lh, long entryId, Object ctx)
     @Override
     public void safeRun() {
         // Remove this entry from the head of the pending queue
-        OpAddEntry firstInQueue = ml.pendingAddEntries.poll();
-        checkArgument(this == firstInQueue);
+        OpAddEntry firstInQueue = ml.pollPendingAddEntry();
+        if (this != firstInQueue) {
+            log.warn("Expected OpAddEntry wasn't first in queue. expected={} actual={}", this, firstInQueue);
+            ManagedLedgerException managedLedgerException =
+                    new ManagedLedgerException("Expected OpAddEntry wasn't first in queue.");
+            // fail both entries
+            try {
+                if (firstInQueue != null) {
+                    firstInQueue.failed(managedLedgerException);
+                }
+            } finally {
+                failed(managedLedgerException);
```

Comment:
Seems the OpAddEntry get failed but does not remove from the pending entries.

Reply:
that's true, but I guess that this PR is doing the right thing by calling `.failed` on the entries to discard them if the execution gets to this point? The previous behavior of using `checkArgument(this == firstInQueue)`  would throw an `IllegalArgumentException` and leak ByteBufs on the discarded OpAddEntrys. Isn't this PR an improvement to that since it blocks the memory leak?
The repro case in #10738 was able to reproduce these errors in the original code amoung others.

@codelipenghui do you expect that some changes would be made to this PR?

Reply:
@lhotari The `checkArgument(..)` call is used in a lot of places in Pulsar. Could it be an issue in other places as well?

Reply:
@devinbost if the arguments get rejected and one of the argument is a [Netty reference counted object](https://netty.io/wiki/reference-counted-objects.html), it would cause a memory leak if the called method is responsible for releasing the object.

Reply:
Regarding the comment from @codelipenghui , I don't remember seeing any logic to handle that case (of removing from pending entries upon failure), so should we just add that logic, or would we want to investigate the cause of why the `get` failed?

Reply:
@devinbost  there is some logic for that. https://github.com/apache/pulsar/blob/6a17b3eed6e18e925b133720482ecdaf246e5ef5/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java#L716-L731
It's a bit hard to understand why the `addOperation` gets added to the `pendingAddEntries` initially. Why doesn't this happen after the state is checked? However I consider handling this a separate issue.

In Netty, the contract of reference counted objects is that ["The general rule of thumb is that the party that accesses a reference-counted object last is also responsible for the destruction of that reference-counted object."](https://netty.io/wiki/reference-counted-objects.html#who-destroys-it). This PR attempts to follow this rule. Therefore I'd suggest that we first process this PR and file separate issues to fix other problems.

Reply:
Looks like it's only removed according to the state, but if the sanity check fails, it's not handled right.